System/Metrics/Prometheus/Internal/State.hs

```diff
@@ -152,6 +161,25 @@ data GroupSampler = forall a. GroupSampler
         -- ^ Metric identifiers and getter functions.
      }
 
+-- | An action to sample an a group of metrics, much like
+-- `GroupSampler`, but the metrics of this group can, at sampling time,
+-- generate any number of values at arbitrary labels. This is the sense
+-- in which these groups are "dynamic".
+--
+-- This type of sampler serve as an escape hatch from the structure
+-- otherwise imposed on metrics by this library. However, users are
+-- expected to ensure that their sampling actions avoid collisions of
+-- metric identifiers; in the case of a collision, we emit neither
+-- warnings nor errors and arbitrarily choose one metric to export. This
+-- is the sense in which these groups are "unchecked".
+data UncheckedDynamicGroupSampler = forall a. UncheckedDynamicGroupSampler
+     { uncheckedDynamicGroupSampleAction :: !(IO a)
+        -- ^ Action to sample the metric group
+     , uncheckedDynamicGroupSamplerMetrics ::
+          !(M.Map Name (Help, a -> M.Map Labels Value))
```

Comment:
I think this is probably not worth doing, but it might be a little cleaner to make this (and similarly in  `groupSamplerMetrics`): `Data.Functor.Compose (M.Map Name) (Help,) (a -> M.Map Labels Value))`, Which would make the application of this function cleaner…  especially if you wanted to use this I guess: https://hackage.haskell.org/package/lens-5.2.2/docs/Control-Lens-Lens.html#v:-63--63-

 but that's also kind of obscure I think ( I didn't know about it).  anyway, just something I noticed, but like I said I think what you have here is probably better

Or maybe you'd like later to add your own helper: `foo ffa a = fmap ($ a) (Compose ffa)`, and use that